docs/proposed/pub-sub.rst

```diff
@@ -0,0 +1,124 @@
+advantagous to send messages to particular clients (or for particular
+topics). A well-known pattern called "Publish/Subscribe" or "PubSub"
+would answer many of these use-cases, but Tahoe-LAFS does not include
+such messaging.
+
+
+Use Cases
+---------
+
+Many of these use-cases are special-cases of a general want:
+
+ - there exists a mutable capability;
+ - one client updates content in this capability;
+ - ..and one or more (different) clients wishes to know when it changes
+
+Thus, a "pub/sub" pattern could be implemented by adding a mechanism
+to notify clients that a particular mutable (to which they possess a
+valid read-capability) has changed. This would maintain many of the
+existing properties of capabilities.
+
+
+Magic Folders Updates
+`````````````````````
+
+Magic Folders in Tahoe currently poll, looking for updates. This
+involves periodically downloading some mutable directory-capabilities
+and determining if anything has changed. This is wasteful of bandwidth
+and server resources. It would nice if a client, Alice, could tell
+other participants in the magic-folder that she's added some
+content. More generally, it would be useful for a client to be able to
+ask a storage-server to tell it whenever a particular (mutable)
+capability has changed. (For the Magic Folders use-case, this would
+mean Alice simply makes changes as now, and relies on the server to
+update interested clients).
+
+
+Grid -> Client Communication
+````````````````````````````
+
+Allowing the operator(s) of a Grid to communicate with clients is
+often desired. For example, notifying users of upcoming maintenance,
+new terms or general news. There is currently no mechanism to do this.
+
+An obvious way to implement this right now would be with a mechanism
+like Magic Folders uses: a mutable capability containing news which
+clients poll (and which the operator(s) can update). This would suffer
+the same polling problems, however.
+
+
+Proposed Solution
+=================
+
+Anything wanting a "broadcast" or "pub-sub" like pattern can implement
+it using the following mechanism:
+
+ - the publisher creates a mutable directory, yielding a write-cap, "W"
+ - anyone wishing to subscribe is given a read-cap, "R"
+   - anyone possessing a valid read-cap "R" (including the publisher)
+     can obviously "invite" further clients
+   - nobody else may subscribe
+ - to add information, the publisher adds a new file to the mutable
+   directory
+ - old information can be removed (or not) as the publisher wishes
+
+Without further changes to Tahoe-LAFS, this can already be achieved:
+clients can "poll" the read-capability and determine if anything new
+exists.
+
+
+Improvements Over Polling
+-------------------------
+
+The above mechanism would scale somewhat poorly as the number of
+"broadcast" capabilities increased. It would be better if the
+storage-server could *tell* clients when a mutable capability has
```

Comment:
Hmm... Maybe I'm misunderstanding how tahoe's mutables work under the hood here, but shouldn't the storage node only be notifying client nodes when a _share_ has updated/changed (and not a full _capability_)? Reasons being: 1) we should not ever want or encourage users/clients to expose the _contents_ of their files to storage nodes (i.e., which, in this case, happens since the storage node receives the cap in the "subscribe" message below) and 2) in an introducerless environment, storage nodes have no knowledge of (or connection to) other the storage nodes that comprise the "grid" (and thus may not be able to tell when a _capability_ has updated unless they're also "subscribed" to the other storage nodes -- which is inefficient -- and/or have some way of monitoring how/where shares are placed on _other_ nodes -- which requires granting excess authority, etc.)...

Reply:
Ah, yes, good catch Chris :) should have had that second coffee before pushing this ;)

Yes, storage-servers only really know about shares. I'll have to look again at the mutable code/specs but it *might* make sense to trigger off "slots". And you're right we should definitely NOT be giving an actual capability to a server :/